### ISSUE TITLE
ESLint Performance Issue

### ISSUE THREAD

**Issue by ashley-hunter** [nrwl/nx#3957]
## Current Behavior

I have a mono repo with 20+ libraries and a few storybooks. Each library contains a single component. 

When I run Eslint (the new builder) on a single library it takes 22 seconds to lint that single library of 1 component.

From my investigation it appears to be the creation of a TS program based on `tsconfig.base.json` which includes files in every app and library, not just the ones in the library I want to lint.

If for example I was to add the following to the .eslintrc.json file in the library folder, so it creates the TS program based on the files solely in that library I get a much more respectable 4 second linting time for that single library:

```json
"parserOptions": {
    "project": "./libs/react/box/tsconfig.*?.json"
}
```

While this workaround works when running the builder my IDE throws warnings so this isn't really a solution, it just demonstrates the issue.

## Expected Behavior

Linting a single library should run fast, and not slow down based on the number of other apps and libraries in the repository.

## Steps to Reproduce

- Clone https://github.com/nrwl/nx-examples/pull/119
- run `yarn nx lint sample --skip-nx-cache`
- In the `libs/sample/.eslintrc.json` file uncomment out the `parserOptions` section
- run the linting command above again
- It should run faster now we are not including all files. Note the speed up is not that significant in this small project, but in a larger repository we are seeing a 5x improvement.

### Failure Logs

N/A

### Environment

```
@nrwl/angular : 10.3.1
 @nrwl/cli : 10.3.0
 @nrwl/cypress : 10.3.1
 @nrwl/eslint-plugin-nx : 10.3.1
 @nrwl/express : Not Found
 @nrwl/jest : 10.3.1
 @nrwl/linter : 10.3.1
 @nrwl/nest : Not Found
 @nrwl/next : Not Found
 @nrwl/node : Not Found
 @nrwl/react : 10.3.1
 @nrwl/schematics : Not Found
 @nrwl/tao : 10.3.0
 @nrwl/web : 10.3.1
 @nrwl/workspace : 10.3.1
 typescript : 4.0.3
```

**Comment by JamesHenry** [comment]
Yeah, this is the manifestation of why I wrote this `TODO` when I last touched the workspace lint utils :) https://github.com/nrwl/nx/blob/992013ca0af7676cae3005b308c7496fa77bba9f/packages/workspace/src/utils/lint.ts#L277

We should be able to address this soon I think @FrozenPandaz - it naturalyl forms part of the overrides update for non-Angular projects

**Comment by ericwooley** [comment]
Is this on the roadmap anywhere? We have a lot of libraries as well, and this is blowing up our CI time.

**Comment by JamesHenry** [comment]
@ericwooley the PR is ready my understanding from Jason is it will be merged in the release after next (11.4 I think)

**Comment by k3nda** [comment]
@JamesHenry  do you have any estimation about next release please?

**Comment by JamesHenry** [comment]
I think @FrozenPandaz is merging the PR today

**Comment by k3nda** [comment]
@JamesHenry so release will be today too? :pleading_face:

**Comment by seyfert** [comment]
Not sure if the PR mentioned above will fix the issue I am experiencing but I have a workspace with almost 70 projects (over 1300 TS files) in an Nx workspace and it takes more than 5 minutes to lint all the projects using `nx run-many --target=lint --all --parallel --maxParallel=4`.

I added caching to each project's lint options so that with incremental linting within projects would work (have to specify a cache-location to prevent projects from clobbering each other):
```
        "lint": {
          "builder": "@nrwl/linter:eslint",
          "options": {
            "cache": true,
            "cacheLocation": "node_modules/.cache/my-app.eslintcache",
            "lintFilePatterns": [
              "apps/my-app/**/*.ts",
              "apps/my-app/**/*.html"
            ]
          }
        },
```
However, linting every project the first time still took 5 minutes.

There seems to be a startup cost to running eslint. As @ashley-hunter mentioned above, linting a single project took about 20 seconds and since using `nx run-many` is essentially running eslint 70 times (although it does run them in parallel). I found that I could lint multiple directories far more efficiently using eslint directly:
```
eslint apps/my-app libs/my-lib1 libs/my-lib2
```
Using the command below to generate a list of paths for all projects defined in angular.json (you will have to install jq) I was able to lint my entire workspace in 40 seconds (without cache). With caching enabled and having a few stale files takes less than 10 seconds. If everything is cached it only takes 3 seconds.
```
jq -r '.projects | to_entries | .[].value.root' angular.json | xargs eslint --cache --cache-location node_modules/.cache/.eslintcache
```
I am sure there are other ways this could be improved (for instance only using eslint on projects that are configured to use @nrwl/linter:eslint) but it is a start.  Hopefully Nx can find a way to handle this more efficiently in the future but at the moment relying on Nx's computational cache doesn't seem to be useful with eslint.

**Comment by JamesHenry** [comment]
Hi @seyfert, please do upgrade to 11.5.0 or above and ensure you have run the migrations to update all your project-level `.eslintrc.json` files.

Having `parserOptions.project` defined at the project level and having no root level fallback would provide a very significant performance increase for a repo of your size, if it has not already been configured that way.

**Comment by JamesHenry** [comment]
I'm going to close this one for now, anyone please feel free to open a new issue as long as you are using the latest Nx and have applied all the migrations.
